src/ReplicatedStorage/Client/InstanceComponents/Misc/OceanDamage.lua

```diff
@@ -7,15 +7,13 @@ local OceanDamage = {}
 OceanDamage.__index = OceanDamage
 
 function OceanDamage.new(character: Model)
-	local humanoid = character:WaitForChild("Humanoid")
-	local humanoidRootPart = character:WaitForChild("HumanoidRootPart")
-
 	local self = setmetatable({}, OceanDamage)
 
+	self._humanoid = character:WaitForChild("Humanoid")
 	self._trove = Trove.new()
 	self._trove:Connect(RunService.Heartbeat, function(delta)
-		if humanoidRootPart.Position.Y > OceanDamageConfig.Height then return end
-		humanoid:TakeDamage(delta * OceanDamageConfig.Damage)
+		if self._humanoid.RootPart.Position.Y > OceanDamageConfig.Height then return end
+		self._humanoid:TakeDamage(delta * OceanDamageConfig.Damage)
```

Comment:
You're forgetting a `nil` check again. Here's how I might write it (I highlighted the important line):
```lua
local humanoid = self._humanoid
local rootPart = humanoid.RootPart
if not rootPart then return end -- The important bit you're missing

if rootPart.Position.Y > OceanDamageConfig.Height then return end
humanoid:TakeDamage(delta * OceanDamageConfig.Damage)
```

The extra variables can be helpful in spotting issues sometimes, and can sometimes actually save time writing code since they're a whole word rather than an expression. Much easier to update, and replace, and read, but understandably it feels like overkill at first, dunno.

Reply:
Yeah. This is entirely correct - localizing some of these variables likely isn't necessary though. It's a mild performance improvement, sure - but I think it's likely a micro-optimization in some cases. That said, what Hexcede said is correct, and the nil check, however done, is necessary.